src/ILLink.RoslynAnalyzer/LocalDataFlow/LocalDataFlowVisitor.cs

```diff
@@ -0,0 +1,186 @@
+			(LocalStateLattice, Context) = (lattice, context);
+
+		public void Transfer (BlockWrapper block, LocalState<TValue> state)
+		{
+			foreach (IOperation operation in block.Block.Operations)
+				Visit (operation, state);
+
+			// Blocks may end with a BranchValue computation. Visit the BranchValue operation after all others.
+			IOperation? branchValueOperation = block.Block.BranchValue;
+			if (branchValueOperation != null) {
+				var branchValue = Visit (branchValueOperation, state);
+
+				// BranchValue may represent a value used in a conditional branch to the ConditionalSuccessor - if so, we are done.
+
+				// If not, the BranchValue represents a return or throw value associated with the FallThroughSuccessor of this block.
+				// (ConditionalSuccessor == null iff ConditionKind == None).
+				if (block.Block.ConditionKind == ControlFlowConditionKind.None) {
+					// This means it's a return value or throw value associated with the fall-through successor.
+
+					// Return statements with return values are represented in the control flow graph as
+					// a branch value operation that computes the return value. The return operation itself is the parent
+					// of the branch value operation.
+
+					// Get the actual "return Foo;" operation (not the branch value operation "Foo").
+					// This should be used as the location of the warning. This is important to provide the right
+					// warning location and because warnings are disambiguated based on the operation.
+					var parentSyntax = branchValueOperation.Syntax.Parent;
+					if (parentSyntax == null)
+						throw new InvalidOperationException ();
+
+					var parentOperation = Context.Compilation.GetSemanticModel (branchValueOperation.Syntax.SyntaxTree).GetOperation (parentSyntax);
+
+					if (parentOperation is IThrowOperation)
+						throw new NotImplementedException ();
+
+					if (parentOperation is not IReturnOperation returnOperation)
+						throw new InvalidOperationException ();
+
+					HandleReturnValue (branchValue, returnOperation);
+				}
+			}
+		}
+
+		public abstract void HandleAssignment (TValue source, TValue target, IOperation operation);
+
+		public abstract void HandleReceiverArgument (TValue receiver, IInvocationOperation operation);
+
+		public abstract void HandleArgument (TValue argument, IArgumentOperation operation);
+
+		// This takes an IOperation rather than an IReturnOperation because the return value
+		// may (must?) come from BranchValue of an operation whose FallThroughSuccessor is the exit block.
+		public abstract void HandleReturnValue (TValue returnValue, IOperation operation);
+
+#pragma warning disable CS8765
+		// Override with a non-nullable return value to prevent some warnings.
+		// This is safe because the interface constraint on TValue ensures that it's not a nullable type.
+		public override TValue Visit (IOperation operation, LocalState<TValue> argument) => operation.Accept (this, argument)!;
```

Comment:
Are you sure the warning here is about `TValue` not `operation`?

Reply:
Great catch, thank you! PTAL and see if this looks right now. Generics with nullable are confusing... 😖